dags/trending-projects-snapshot-dag.py

```diff
@@ -0,0 +1,192 @@
+    "params": dag_params,
+}
+
+
+@dataclass
+class SnapshotMetrics:
+    """
+    Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        project_id: The ID of the Project
+        n_unique_users: The number of unique users who downloaded from the Project
+        last_download_date: The date of the last download from the Project
+        estimated_project_size_in_gib: The current total size of the Project in GiB
+    """
+
+    project_id: float
+    n_unique_users: int
+    last_download_date: int
+    estimated_project_size_in_gib: float
+
+
+@dag(**dag_config)
+def trending_projects_snapshot() -> None:
+    """
+    This DAG executes a query on Snowflake to retrieve information about trending public projects and 
+    reports the results to a Synapse table.
+
+    The main steps performed in this DAG are:
+    1. Query Snowflake to identify public projects, their file sizes, and recent download activities.
+    2. Aggregate the results to determine num. of unique users, last download date, and total data size.
+    3. Format the results into a list of SnapshotsMetric objects.
+
+    The Snowflake connection and Synapse connection IDs, as well as the time window for recent downloads,
+    are provided via DAG parameters.
+    """
+
+    @task
+    def get_trending_project_snapshot(**context) -> List[SnapshotMetrics]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'project'
+                    AND IS_PUBLIC
+                ),
+                RECENT_DOWNLOADS AS (
+                    SELECT PROJECT_ID, RECORD_DATE, USER_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILEDOWNLOAD
+                    WHERE 1=1
+                    AND DATE_TRUNC('MONTH', RECORD_DATE) = DATE_TRUNC('MONTH', DATE('{context["params"]["month_to_run"]}'))
+                    AND STACK = 'prod'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
+                ),
+                TOP_10_PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID, COUNT(DISTINCT USER_ID) AS N_UNIQUE_USERS
+                    FROM RECENT_DOWNLOADS
+                    GROUP BY PROJECT_ID
+                    ORDER BY N_UNIQUE_USERS DESC
+                    LIMIT 10
+                ),
+                LATEST_FILE_HANDLES AS (
+                    SELECT ID, PROJECT_ID, FILE_HANDLE_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND DATE_TRUNC('MONTH', CHANGE_TIMESTAMP) <= DATE_TRUNC('MONTH', DATE('{context["params"]["month_to_run"]}'))
```

Comment:
This will be an issue because sometimes the CHANGE_TIMESTAMP will be greater than the CREATEDON date. Let's say an entity was created in Feb but updated in April, by doing this query, when you do the calculation of data sizes for Feb, you're going to leave this entity out.

Reply:
Got it, I've updated it so that as long as the file was created <= the `month_to_run`, it will be used in the calculation of the total Project size.